Move functional tests to new workflow, add ability to run manually

danielhfrank opened the pull request:
# Background

I spoke with @bnaul earlier about having the ability to run functional tests manually in the model repo, and observed @mark-idleman [struggling](https://github.com/replicahq/graphhopper/runs/3295091696) to repro issues running functional tests in this repo. I even had my own functionality I wanted to work on, but got sucked into this instead. Basically, how can we have functionality that we mostly want to run only on merges to master, but that we can still exercise ourselves to ensure that it works as we develop on it?

# Description

tl;dr I added the ability to run the functional tests on a branch, by running `gh workflow run "Functional Test" --ref my-branch`

The way that I elected to implement this was to break the functional tests out into their own workflow, which itself can be triggered by the completion of the original workflow: https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run . I then also added the ability to trigger the workflow manually. Finally, I put in place some conditionals to restrict when the two steps run. The step that actually runs functional tests will execute either on a manual trigger, or when the upstream workflow succeeds on the main branch. The tagging step will only execute in the latter case.

I should admit, the more complicated these conditions get, the more fraught it becomes to ensure that they will all behave as intended (see commit history here). My hope is that we don't get too crazy with the logic like this that we implement, but rather establish a couple of common patterns that we use.

# Testing
I was able to run just the functional tests here: https://github.com/replicahq/graphhopper/actions/runs/1128977093 . Note that I didn't need to comment anything out, although I did need to wait for the test image to build.

# cc
@replicahq/router

Commit:
Start on a manually-triggerable functional test

Commit:
Make sure steps are a list

Commit:
Try to refresh

Commit:
Make it only run if I ask it to

Commit:
Move conditional

Commit:
Single quotes

Commit:
thrash

Commit:
Reduce conditional, wtf

Commit:
other conditional

Commit:
Try with less escaping

Commit:
More complicated condition

Commit:
Move functional test steps to their own workflow

Commit:
Restrict tagging to only on workflow_run trigger

JacobHayes commented on .github/workflows/functional-test.yaml:
```diff
@@ -0,0 +1,128 @@
+  update-tags:
+    runs-on: ubuntu-latest
+    needs: functional-tests
+    if: github.ref == 'refs/heads/original-direction' && github.event.workflow_run.conclusion == 'success'
```
Maybe add a small note/comment explicitly saying we don't want to run on `workflow_dispatch`?

JacobHayes commented on .github/workflows/functional-test.yaml:
```diff
@@ -0,0 +1,128 @@
+jobs:
+  functional-tests:
+    name: "Build server image and run functional tests"
+    if: github.event_name == 'workflow_dispatch' || (github.ref == 'refs/heads/original-direction' && github.event.workflow_run.conclusion == 'success' )
```
nit:
```suggestion
    if: github.event_name == 'workflow_dispatch' || (github.ref == 'refs/heads/original-direction' && github.event.workflow_run.conclusion == 'success')
```

JacobHayes reviewed (approved):
Makes sense to me, thanks for tackling!

Commit:
Comment and whitespace

danielhfrank commented:
I've added some comments about usage and cleaned up the style stuff that @JacobHayes identified. There were also some significant merge conflicts to address, as I intentionally branched this work off of an older revision where functional tests were passing (they are currently broken on master). I just rebased, and confirmed that the current revision of functional tests [fails](https://github.com/replicahq/graphhopper/runs/3343298503?check_suite_focus=true) in the same way as the [revision on master](https://github.com/replicahq/graphhopper/runs/3295091696?check_suite_focus=true). The idea here is that being able to run the workflow manually will allow the router team to iterate on fixing the functional test. Going to merge now so that they can do that.

mark-idleman commented:
I'll pick up "fixing the functional test: part 2" now that we can easily test it on non-`original-direction` commits 😁
